### ISSUE TITLE
some problems about running repo

### ISSUE THREAD

**Issue by ChenJiaJiaJia** [PRBonn/ContMAV#2]
I am following your excellent work. But The process was very unsmooth. Please help me .
-------
1. is it necessary to run below code in "src/datasets/cityscapes/README.md"?
```
    CITYSCAPES_DOWNLOAD_DIR="/home/deltron/workspace/cjj/ContMAV/cityscapes_downloads"
    # labels
    csDownload gtFine_trainvaltest.zip -d $CITYSCAPES_DOWNLOAD_DIR    # -> 241MB
    # rgb images
    csDownload leftImg8bit_trainvaltest.zip -d $CITYSCAPES_DOWNLOAD_DIR     # -> 11GB
    # disparity images (only upon request)
    csDownload disparity_trainvaltest.zip -d $CITYSCAPES_DOWNLOAD_DIR     # -> 3.5GB  
   
    # unzip files
    find $CITYSCAPES_DOWNLOAD_DIR -name '*.zip' -exec unzip -o {} -d $CITYSCAPES_DOWNLOAD_DIR \;

    python prepare_dataset.py ../../../datasets/cityscapes $CITYSCAPES_DOWNLOAD_DIR
```
when ran above code , I met some issues:
- lack of "camera" dir
  - I fixed it through downloading the camera part of cityscapes dataset
- `assert all(l == identifier_lists[0] for l in identifier_lists[1:])` of `prepare_dataset.py` interrupt the  process.  Because there is no "valid" part in datasets.
  - I commented this sentence later.

in the end , it output the results in `datasets/cityscapes`. (generate the data according to expected format)

-------

2. However , after  completing the step 1 (above) , I still encountered some problems regarding the dataset when running the train.py code .
I did some changes so that the process can run.
- I revise 
```
self.images_path = os.path.join(data_dir, "leftImg8bit", split)
self.labels_path = os.path.join(data_dir, "gtFine", split)
```
to 
```
self.images_path = os.path.join(data_dir, split, "rgb")
self.labels_path = os.path.join(data_dir, split, "labels_19")
```

- and , I revise
```
for filename in glob.iglob(
    self.labels_path + "/**/*labelTrainIds.png", recursive=True
):
    self.labels.append(filename)
```
to
```
for filename in glob.iglob(
    self.labels_path + "/**/*labelIds.png", recursive=True
):
    self.labels.append(filename)
```

-------

3.  I encounted below issus:
```
Traceback (most recent call last):
  File "/home/deltron/workspace/cjj/ContMAV/train.py", line 624, in <module>
    train_main()
  File "/home/deltron/workspace/cjj/ContMAV/train.py", line 176, in train_main
    mean, var = train_one_epoch(
  File "/home/deltron/workspace/cjj/ContMAV/train.py", line 283, in train_one_epoch
    losses = loss_function_train(pred_scales, cw_target)
  File "/home/deltron/miniconda3/envs/openworld/lib/python3.10/site-packages/torch/nn/modules/module.py", line 1190, in _call_impl
    return forward_call(*input, **kwargs)
  File "/home/deltron/workspace/cjj/ContMAV/src/utils.py", line 277, in forward
    number_of_pixels_per_class[1:] * self.weight
RuntimeError: The size of tensor a (20) must match the size of tensor b (19) at non-singleton dimension 0
```
I revise 
```
number_of_pixels_per_class = torch.bincount(
    targets.flatten().type(self.dtype), minlength=self.num_classes
)
```
to
```
number_of_pixels_per_class = torch.bincount(
    targets_m.flatten().type(self.dtype), minlength=self.num_classes
)
```
it is in `utils.py`. 
is it right?

-------

4. (the biggest problem)
I can't figure out what causes it to show different errors when I execute it at different times. (I have set some global random seed).

one of perfermance is occuring `Floating point exception (core dumped)` when run code . but when debug code (using vscode), the process is interrupted directly without any except.

another perfermance is:
after executing `loss_all = self.ce_loss(inputs, targets_m_long)` in `utils.py` , if I execute any code about tensor, an error will be reported. 
one example shown below:
```
Traceback (most recent call last):
  File "/home/deltron/miniconda3/envs/openworld/lib/python3.10/runpy.py", line 196, in _run_module_as_main
    return _run_code(code, main_globals, None,
  File "/home/deltron/miniconda3/envs/openworld/lib/python3.10/runpy.py", line 86, in _run_code
    exec(code, run_globals)
  File "/home/deltron/.vscode-server/extensions/ms-python.debugpy-2024.6.0-linux-x64/bundled/libs/debugpy/adapter/../../debugpy/launcher/../../debugpy/__main__.py", line 39, in <module>
    cli.main()
  File "/home/deltron/.vscode-server/extensions/ms-python.debugpy-2024.6.0-linux-x64/bundled/libs/debugpy/adapter/../../debugpy/launcher/../../debugpy/../debugpy/server/cli.py", line 430, in main
    run()
  File "/home/deltron/.vscode-server/extensions/ms-python.debugpy-2024.6.0-linux-x64/bundled/libs/debugpy/adapter/../../debugpy/launcher/../../debugpy/../debugpy/server/cli.py", line 284, in run_file
    runpy.run_path(target, run_name="__main__")
  File "/home/deltron/.vscode-server/extensions/ms-python.debugpy-2024.6.0-linux-x64/bundled/libs/debugpy/_vendored/pydevd/_pydevd_bundle/pydevd_runpy.py", line 321, in run_path
    return _run_module_code(code, init_globals, run_name,
  File "/home/deltron/.vscode-server/extensions/ms-python.debugpy-2024.6.0-linux-x64/bundled/libs/debugpy/_vendored/pydevd/_pydevd_bundle/pydevd_runpy.py", line 135, in _run_module_code
    _run_code(code, mod_globals, init_globals,
  File "/home/deltron/.vscode-server/extensions/ms-python.debugpy-2024.6.0-linux-x64/bundled/libs/debugpy/_vendored/pydevd/_pydevd_bundle/pydevd_runpy.py", line 124, in _run_code
    exec(code, run_globals)
  File "/home/deltron/workspace/cjj/ContMAV/train.py", line 613, in <module>
    train_main()
  File "/home/deltron/workspace/cjj/ContMAV/train.py", line 175, in train_main
    mean, var = train_one_epoch(
  File "/home/deltron/workspace/cjj/ContMAV/train.py", line 281, in train_one_epoch
    losses = loss_function_train(pred_scales, cw_target)
  File "/home/deltron/miniconda3/envs/openworld/lib/python3.10/site-packages/torch/nn/modules/module.py", line 1190, in _call_impl
    return forward_call(*input, **kwargs)
  File "/home/deltron/workspace/cjj/ContMAV/src/utils.py", line 249, in forward
    number_of_pixels_per_class = torch.bincount(
RuntimeError: CUDA error: an illegal memory access was encountered
CUDA kernel errors might be asynchronously reported at some other API call,so the stacktrace below might be incorrect.
For debugging consider passing CUDA_LAUNCH_BLOCKING=1.
```

**Comment by ChenJiaJiaJia** [comment]
I encounted a new problem :
```
Train Epoch:   3 [1192/2975 ( 40.1%)]   lr_0: 1.2213e-06   Loss: 1.465581  [0.12s every    2 data]
Train Epoch:   3 [1194/2975 ( 40.1%)]   lr_0: 1.2213e-06   Loss: 1.090607  [0.14s every    2 data]
Train Epoch:   3 [1196/2975 ( 40.2%)]   lr_0: 1.2213e-06   Loss: 0.911586  [0.12s every    2 data]
Train Epoch:   3 [1198/2975 ( 40.3%)]   lr_0: 1.2213e-06   Loss: 1.109312  [0.13s every    2 data]
> /home/deltron/workspace/cjj/ContMAV/train.py(324)train_one_epoch()
    323             import ipdb;ipdb.set_trace()  # fmt: skip
--> 324             raise ValueError("Loss is None")
    325 

ipdb> total_loss
array(nan)
ipdb> loss_con
array(0., dtype=float32)
ipdb> loss_ows
array(0., dtype=float32)
ipdb> loss_objectosphere
array(0., dtype=float32)
ipdb> loss_segmentation
array(nan)
```

how to fix it

**Comment by magistri** [comment]
Hi!
It's hard to tell what's going on without more details, it seems to be a problem of the cross-entropy loss. Can you visualize/print the target tensor for the cross entropy loss? Are the input images properly loaded?

**Comment by Runzhong-Zhang** [comment]
Hi Chenjiajiajia and magistri:

I encounter the same issue as in forward:
    number_of_pixels_per_class = torch.bincount(
RuntimeError: CUDA error: an illegal memory access was encountered

Would you kindly help out if there's any solution for it?

**Comment by magistri** [comment]
This is likely a torch issue. Could you run CUDA_LAUNCH_BLOCKING=1 python train.py and post the error again?

**Comment by Runzhong-Zhang** [comment]
Sure, the full log is listed below:

(openworld) (base) runzhong@c5773:~/repo/3-2/V0-ContMAV$ CUDA_LAUNCH_BLOCKING=1 python train.py
Notice: adapting learning rate to 2.5e-05 because provided batch size differs from default batch size of 8.
train
val
test
Notice: argument --channels_decoder is ignored when --decoder_chanels_mode decreasing is set.
Notice: for the context module the learned upsampling is not possible as the feature maps are not upscaled by the factor 2. We will use nearest neighbor instead.
Device: cuda:0



Total params: 48.33M



Applied He init.
Using /home/runzhong/repo/3-2/V0-ContMAV/src/datasets/cityscapes/weighting_linear_1+19_val.pickle as class weighting
Using Adam as optimizer


=========================================================================


/home/runzhong/miniconda3/envs/openworld/lib/python3.10/site-packages/torch/optim/lr_scheduler.py:138: UserWarning: Detected call of `lr_scheduler.step()` before `optimizer.step()`. In PyTorch 1.1.0 and later, you should call them in the opposite order: `optimizer.step()` before `lr_scheduler.step()`.  Failure to do this will result in PyTorch skipping the first value of the learning rate schedule. See more details at https://pytorch.org/docs/stable/optim.html#how-to-adjust-learning-rate
  warnings.warn("Detected call of `lr_scheduler.step()` before `optimizer.step()`. "
/home/runzhong/miniconda3/envs/openworld/lib/python3.10/site-packages/torch/optim/lr_scheduler.py:163: UserWarning: The epoch parameter in `scheduler.step()` was not necessary and is being deprecated where possible. Please use `scheduler.step()` to step the scheduler. During the deprecation, if epoch is different from None, the closed form is used instead of the new chainable form, where available. Please open an issue if you are unable to replicate your use case: https://github.com/pytorch/pytorch/issues/new/choose.
  warnings.warn(EPOCH_DEPRECATION_WARNING, UserWarning)
/home/runzhong/miniconda3/envs/openworld/lib/python3.10/site-packages/torch/optim/lr_scheduler.py:1679: UserWarning: To get the last learning rate computed by the scheduler, please use `get_last_lr()`.
  warnings.warn("To get the last learning rate computed by the scheduler, "
Train Epoch:   0 [   2/  40 (  5.0%)]   lr_0:  1e-06   Loss: 3.506484  [2.33s every    2 data]
Train Epoch:   0 [   4/  40 ( 10.0%)]   lr_0:  1e-06   Loss: 3.059636  [0.14s every    2 data]
Traceback (most recent call last):
  File "/home/runzhong/repo/3-2/V0-ContMAV/train.py", line 592, in <module>
    train_main()
  File "/home/runzhong/repo/3-2/V0-ContMAV/train.py", line 165, in train_main
    mean, var = train_one_epoch(
  File "/home/runzhong/repo/3-2/V0-ContMAV/train.py", line 265, in train_one_epoch
    losses = loss_function_train(pred_scales, cw_target)
  File "/home/runzhong/miniconda3/envs/openworld/lib/python3.10/site-packages/torch/nn/modules/module.py", line 1190, in _call_impl
    return forward_call(*input, **kwargs)
  File "/home/runzhong/repo/3-2/V0-ContMAV/src/utils.py", line 238, in forward
    number_of_pixels_per_class = torch.bincount(
RuntimeError: CUDA error: an illegal memory access was encountered

Many thanks for your kind help!

**Comment by Runzhong-Zhang** [comment]
When running the same CUDA_LAUNCH_BLOCKING=1 python train.py command, it also randomly returns the "Floating point exception (core dumped)"  bug as follows:

Train Epoch:   0 [   2/  40 (  5.0%)]   lr_0:  1e-06   Loss: 2.490788  [2.39s every    2 data]
Train Epoch:   0 [   4/  40 ( 10.0%)]   lr_0:  1e-06   Loss: 2.539077  [0.14s every    2 data]
Train Epoch:   0 [   6/  40 ( 15.0%)]   lr_0:  1e-06   Loss: 2.812910  [0.13s every    2 data]
Train Epoch:   0 [   8/  40 ( 20.0%)]   lr_0:  1e-06   Loss: 2.583890  [0.13s every    2 data]
Train Epoch:   0 [  10/  40 ( 25.0%)]   lr_0:  1e-06   Loss: 2.704608  [0.13s every    2 data]
Train Epoch:   0 [  12/  40 ( 30.0%)]   lr_0:  1e-06   Loss: 2.964348  [0.13s every    2 data]
Train Epoch:   0 [  14/  40 ( 35.0%)]   lr_0:  1e-06   Loss: 2.762590  [0.13s every    2 data]
Train Epoch:   0 [  16/  40 ( 40.0%)]   lr_0:  1e-06   Loss: 2.879232  [0.13s every    2 data]
Train Epoch:   0 [  18/  40 ( 45.0%)]   lr_0:  1e-06   Loss: 2.957944  [0.13s every    2 data]
Train Epoch:   0 [  20/  40 ( 50.0%)]   lr_0:  1e-06   Loss: 3.199244  [0.13s every    2 data]
Train Epoch:   0 [  22/  40 ( 55.0%)]   lr_0:  1e-06   Loss: 2.903895  [0.13s every    2 data]
Train Epoch:   0 [  24/  40 ( 60.0%)]   lr_0:  1e-06   Loss: 2.319134  [0.13s every    2 data]
Train Epoch:   0 [  26/  40 ( 65.0%)]   lr_0:  1e-06   Loss: 2.580545  [0.13s every    2 data]
Train Epoch:   0 [  28/  40 ( 70.0%)]   lr_0:  1e-06   Loss: 2.236407  [0.13s every    2 data]
Floating point exception (core dumped)

To summarize, I encountered the same issue reported in the "ChenJiaJia 4. (the biggest problem)" section. Thanks again for your great help!

**Comment by magistri** [comment]
Thanks! I'd guess both errors can be the results of an empty tensor. I'm not entirely sure how this can happen tho. 
Can you catch the RuntimeError on number_of_pixels_per_class and see if there's really an empty tensor?

**Comment by huynhtruc0309** [comment]
Hi everyone, I am facing the same error. Does anyone get a solution for this?

**Comment by 78748547** [comment]
> ## I am following your excellent work. But The process was very unsmooth. Please help me .
> 1. is it necessary to run below code in "src/datasets/cityscapes/README.md"?
> 
> ```
>     CITYSCAPES_DOWNLOAD_DIR="/home/deltron/workspace/cjj/ContMAV/cityscapes_downloads"
>     # labels
>     csDownload gtFine_trainvaltest.zip -d $CITYSCAPES_DOWNLOAD_DIR    # -> 241MB
>     # rgb images
>     csDownload leftImg8bit_trainvaltest.zip -d $CITYSCAPES_DOWNLOAD_DIR     # -> 11GB
>     # disparity images (only upon request)
>     csDownload disparity_trainvaltest.zip -d $CITYSCAPES_DOWNLOAD_DIR     # -> 3.5GB  
>    
>     # unzip files
>     find $CITYSCAPES_DOWNLOAD_DIR -name '*.zip' -exec unzip -o {} -d $CITYSCAPES_DOWNLOAD_DIR \;
> 
>     python prepare_dataset.py ../../../datasets/cityscapes $CITYSCAPES_DOWNLOAD_DIR
> ```
> 
> when ran above code , I met some issues:
> 
> * lack of "camera" dir
>   
>   * I fixed it through downloading the camera part of cityscapes dataset
> * `assert all(l == identifier_lists[0] for l in identifier_lists[1:])` of `prepare_dataset.py` interrupt the  process.  Because there is no "valid" part in datasets.
>   
>   * I commented this sentence later.
> 
> in the end , it output the results in `datasets/cityscapes`. (generate the data according to expected format)
> 
> 2. However , after  completing the step 1 (above) , I still encountered some problems regarding the dataset when running the train.py code .
>    I did some changes so that the process can run.
> 
> * I revise
> 
> ```
> self.images_path = os.path.join(data_dir, "leftImg8bit", split)
> self.labels_path = os.path.join(data_dir, "gtFine", split)
> ```
> 
> to
> 
> ```
> self.images_path = os.path.join(data_dir, split, "rgb")
> self.labels_path = os.path.join(data_dir, split, "labels_19")
> ```
> 
> * and , I revise
> 
> ```
> for filename in glob.iglob(
>     self.labels_path + "/**/*labelTrainIds.png", recursive=True
> ):
>     self.labels.append(filename)
> ```
> 
> to
> 
> ```
> for filename in glob.iglob(
>     self.labels_path + "/**/*labelIds.png", recursive=True
> ):
>     self.labels.append(filename)
> ```
> 
> 3. I encounted below issus:
> 
> ```
> Traceback (most recent call last):
>   File "/home/deltron/workspace/cjj/ContMAV/train.py", line 624, in <module>
>     train_main()
>   File "/home/deltron/workspace/cjj/ContMAV/train.py", line 176, in train_main
>     mean, var = train_one_epoch(
>   File "/home/deltron/workspace/cjj/ContMAV/train.py", line 283, in train_one_epoch
>     losses = loss_function_train(pred_scales, cw_target)
>   File "/home/deltron/miniconda3/envs/openworld/lib/python3.10/site-packages/torch/nn/modules/module.py", line 1190, in _call_impl
>     return forward_call(*input, **kwargs)
>   File "/home/deltron/workspace/cjj/ContMAV/src/utils.py", line 277, in forward
>     number_of_pixels_per_class[1:] * self.weight
> RuntimeError: The size of tensor a (20) must match the size of tensor b (19) at non-singleton dimension 0
> ```
> 
> I revise
> 
> ```
> number_of_pixels_per_class = torch.bincount(
>     targets.flatten().type(self.dtype), minlength=self.num_classes
> )
> ```
> 
> to
> 
> ```
> number_of_pixels_per_class = torch.bincount(
>     targets_m.flatten().type(self.dtype), minlength=self.num_classes
> )
> ```
> 
> it is in `utils.py`. is it right?
> 
> 4. (the biggest problem)
>    I can't figure out what causes it to show different errors when I execute it at different times. (I have set some global random seed).
> 
> one of perfermance is occuring `Floating point exception (core dumped)` when run code . but when debug code (using vscode), the process is interrupted directly without any except.
> 
> another perfermance is: after executing `loss_all = self.ce_loss(inputs, targets_m_long)` in `utils.py` , if I execute any code about tensor, an error will be reported. one example shown below:
> 
> ```
> Traceback (most recent call last):
>   File "/home/deltron/miniconda3/envs/openworld/lib/python3.10/runpy.py", line 196, in _run_module_as_main
>     return _run_code(code, main_globals, None,
>   File "/home/deltron/miniconda3/envs/openworld/lib/python3.10/runpy.py", line 86, in _run_code
>     exec(code, run_globals)
>   File "/home/deltron/.vscode-server/extensions/ms-python.debugpy-2024.6.0-linux-x64/bundled/libs/debugpy/adapter/../../debugpy/launcher/../../debugpy/__main__.py", line 39, in <module>
>     cli.main()
>   File "/home/deltron/.vscode-server/extensions/ms-python.debugpy-2024.6.0-linux-x64/bundled/libs/debugpy/adapter/../../debugpy/launcher/../../debugpy/../debugpy/server/cli.py", line 430, in main
>     run()
>   File "/home/deltron/.vscode-server/extensions/ms-python.debugpy-2024.6.0-linux-x64/bundled/libs/debugpy/adapter/../../debugpy/launcher/../../debugpy/../debugpy/server/cli.py", line 284, in run_file
>     runpy.run_path(target, run_name="__main__")
>   File "/home/deltron/.vscode-server/extensions/ms-python.debugpy-2024.6.0-linux-x64/bundled/libs/debugpy/_vendored/pydevd/_pydevd_bundle/pydevd_runpy.py", line 321, in run_path
>     return _run_module_code(code, init_globals, run_name,
>   File "/home/deltron/.vscode-server/extensions/ms-python.debugpy-2024.6.0-linux-x64/bundled/libs/debugpy/_vendored/pydevd/_pydevd_bundle/pydevd_runpy.py", line 135, in _run_module_code
>     _run_code(code, mod_globals, init_globals,
>   File "/home/deltron/.vscode-server/extensions/ms-python.debugpy-2024.6.0-linux-x64/bundled/libs/debugpy/_vendored/pydevd/_pydevd_bundle/pydevd_runpy.py", line 124, in _run_code
>     exec(code, run_globals)
>   File "/home/deltron/workspace/cjj/ContMAV/train.py", line 613, in <module>
>     train_main()
>   File "/home/deltron/workspace/cjj/ContMAV/train.py", line 175, in train_main
>     mean, var = train_one_epoch(
>   File "/home/deltron/workspace/cjj/ContMAV/train.py", line 281, in train_one_epoch
>     losses = loss_function_train(pred_scales, cw_target)
>   File "/home/deltron/miniconda3/envs/openworld/lib/python3.10/site-packages/torch/nn/modules/module.py", line 1190, in _call_impl
>     return forward_call(*input, **kwargs)
>   File "/home/deltron/workspace/cjj/ContMAV/src/utils.py", line 249, in forward
>     number_of_pixels_per_class = torch.bincount(
> RuntimeError: CUDA error: an illegal memory access was encountered
> CUDA kernel errors might be asynchronously reported at some other API call,so the stacktrace below might be incorrect.
> For debugging consider passing CUDA_LAUNCH_BLOCKING=1.
> ```


Hello, may I ask how you finally resolved the issue of "* `assert all(l == identifier_lists[0] for l in identifier_lists[1:])` of `prepare_dataset.py` interrupt the  process.  Because there is no "valid" part in datasets."? thank you

**Comment by gitfrahts** [comment]
@ChenJiaJiaJia 
Hello, may I ask how you finally resolved the issue of "* of interrupt the process. Because there is no "valid" part in datasets."? thank youassert all(l == identifier_lists[0] for l in identifier_lists[1:])prepare_dataset.py

**Comment by gitfrahts** [comment]
你解决这个问题了吗？
@78748547

**Comment by AlessandroSciarrillo** [comment]
I’m experiencing the same issues as @ChenJiaJiaJia, and I believe this issue is not resolved. Could someone from the authors provide an update?

**Comment by 78748547** [comment]
> 你解决这个问题了吗？ [@78748547](https://github.com/78748547)

Sorry, I didn't solve the problem. I think there is a problem with this code. I recommend referring to other works related to CLIP
